storage_scrubber/src/checks.rs

```diff
@@ -114,16 +126,40 @@ pub(crate) fn branch_cleanup_and_check_errors(
                         }
 
                         if !tenant_objects.check_ref(id.timeline_id, &layer, &metadata) {
-                            // FIXME: this will emit false positives if an index was
-                            // uploaded concurrently with our scan.  To make this check
-                            // correct, we need to try sending a HEAD request for the
-                            // layer we think is missing.
-                            result.errors.push(format!(
-                                "index_part.json contains a layer {}{} (shard {}) that is not present in remote storage",
+                            let timeline_root = root_target.timeline_root(id);
+                            let remote_layer_path = format!(
+                                "{}{}{}",
+                                timeline_root.prefix_in_bucket,
                                 layer,
-                                metadata.generation.get_suffix(),
-                                metadata.shard
-                            ))
+                                metadata.generation.get_suffix()
+                            );
+
+                            // HEAD request used here to address a race condition  when an index was uploaded concurrently
+                            // with our scan. We check if the object is uploaded to S3 after taking the listing snapshot.
+                            let response = s3_client
```

Comment:
does not seem compatible with Azure but fine for now

Reply:
next step: migrate this to use `GenericRemoteStorage` instead.

Reply:
probably part of #7547. I can help out if needed.